File: backend/engine/pie_calculator.py
"""
Principle of Inclusion–Exclusion (PIE) Engine (Python Reference Implementation)
For Experiential Learning Project — B.Tech CSE (AI-Driven DevOps), JAIN University
Author: Ameya

Mathematical Formulation:
For n sets A_1, A_2, ..., A_n:
|⋃_{i=1}^n A_i| = ∑_{k=1}^n (-1)^{k-1} ∑_{1 ≤ i_1 < ... < i_k ≤ n} |⋂_{j=1}^k A_{i_j}|
"""

from itertools import combinations
from typing import List, Dict, Any

class PIECalculator:
    @staticmethod
    def get_overlap_occupancy(subset: tuple, overlaps: List[Dict[str, Any]]) -> int:
        """Find shared occupancy for an exact subset of zone IDs."""
        sorted_subset = sorted(list(subset))
        for o in overlaps:
            sorted_o = sorted(o.get('zone_ids', []))
            if sorted_o == sorted_subset:
                return int(o.get('shared_occupancy', 0))
        return 0

    @classmethod
    def calculate(cls, zones: List[Dict[str, Any]], overlaps: List[Dict[str, Any]] = None) -> Dict[str, Any]:
        """
        Compute true evacuation occupancy using the Principle of Inclusion-Exclusion.
        
        Args:
            zones: List of zone dictionaries [{'id': 'zone_a', 'name': 'Lab A', 'current_occupancy': 40}, ...]
            overlaps: List of overlaps [{'id': 'ov_1', 'zone_ids': ['zone_a', 'zone_b'], 'shared_occupancy': 15}, ...]
            
        Returns:
            Dictionary with rawSum, correctedOccupancy, overcountingError, and breakdown.
        """
        if not zones:
            return {
                'raw_sum': 0,
                'corrected_occupancy': 0,
                'overcounting_error': 0,
                'error_percentage': 0.0,
                'formula': 'Total = 0'
            }

        if overlaps is None:
            overlaps = []

        zone_map = {z['id']: z for z in zones}
        zone_ids = [z['id'] for z in zones]
        n = len(zones)

        # 1. Level 1: Individual sets (Raw Sum)
        raw_sum = sum(int(z.get('current_occupancy', 0)) for z in zones)
        corrected_occupancy = raw_sum

        formula_parts = [" + ".join(f"|{z.get('name', z['id'])}|" for z in zones)]
        level_details = []

        # 2. Levels 2 through n: Alternating intersections
        for k in range(2, n + 1):
            sign = (-1) ** (k - 1)
            sign_str = '+' if sign > 0 else '−'
            combos = list(combinations(zone_ids, k))
            level_items = []
            level_total = 0

            for combo in combos:
                shared = cls.get_overlap_occupancy(combo, overlaps)
                if shared > 0:
                    names = [zone_map[zid].get('name', zid) for zid in combo]
                    label = f"|{' ∩ '.join(names)}|"
                    level_total += shared
                    level_items.append((label, shared))

            if level_items:
                corrected_occupancy += sign * level_total
                joined_labels = " + ".join(f"{lbl}" for lbl, _ in level_items)
                formula_parts.append(f"{sign_str} ({joined_labels})")
                level_details.append({
                    'k': k,
                    'sign': sign_str,
                    'total': level_total,
                    'items': level_items
                })

        corrected_occupancy = max(0, corrected_occupancy)
        overcounting_error = raw_sum - corrected_occupancy
        error_percentage = round((overcounting_error / raw_sum * 100), 1) if raw_sum > 0 else 0.0

        return {
            'raw_sum': raw_sum,
            'corrected_occupancy': corrected_occupancy,
            'overcounting_error': overcounting_error,
            'error_percentage': error_percentage,
            'formula': " ".join(formula_parts),
            'level_details': level_details
        }

if __name__ == '__main__':
    # Demonstration / Unit Verification
    test_zones = [
        {'id': 'A', 'name': 'Room A', 'current_occupancy': 50},
        {'id': 'B', 'name': 'Room B', 'current_occupancy': 40},
        {'id': 'C', 'name': 'Corridor C', 'current_occupancy': 35}
    ]
    test_overlaps = [
        {'zone_ids': ['A', 'B'], 'shared_occupancy': 15},
        {'zone_ids': ['A', 'C'], 'shared_occupancy': 10},
        {'zone_ids': ['B', 'C'], 'shared_occupancy': 12},
        {'zone_ids': ['A', 'B', 'C'], 'shared_occupancy': 5}
    ]
    result = PIECalculator.calculate(test_zones, test_overlaps)
    print("--- Principle of Inclusion-Exclusion Verification ---")
    print(f"Raw Sum: {result['raw_sum']}")
    print(f"PIE Corrected: {result['corrected_occupancy']} (Expected: 93)")
    print(f"Overcounting Prevented: {result['overcounting_error']} ({result['error_percentage']}%)")
    print(f"Formula: {result['formula']}")

File: backend/engine/pie_calculator.js
/**
 * Principle of Inclusion–Exclusion (PIE) Engine
 * 
 * Mathematical Formulation:
 * For n sets A_1, A_2, ..., A_n:
 * |⋃_{i=1}^n A_i| = ∑_{k=1}^n (-1)^{k-1} ∑_{1 ≤ i_1 < ... < i_k ≤ n} |⋂_{j=1}^k A_{i_j}|
 * 
 * For 2 zones: |A ∪ B| = |A| + |B| - |A ∩ B|
 * For 3 zones: |A ∪ B ∪ C| = |A| + |B| + |C| - (|A ∩ B| + |A ∩ C| + |B ∩ C|) + |A ∩ B ∩ C|
 */

class PIECalculator {
  /**
   * Helper to generate all combinations of array elements of length k
   */
  static getCombinations(arr, k) {
    if (k === 0) return [[]];
    if (arr.length === 0) return [];
    const head = arr[0];
    const tail = arr.slice(1);
    const withHead = PIECalculator.getCombinations(tail, k - 1).map(c => [head, ...c]);
    const withoutHead = PIECalculator.getCombinations(tail, k);
    return [...withHead, ...withoutHead];
  }

  /**
   * Look up shared occupancy for a given subset of zone IDs from overlaps definition.
   * @param {string[]} subset - array of zone IDs e.g. ['zone_a', 'zone_b']
   * @param {Array} overlaps - list of overlap objects { id, zone_ids: [...], shared_occupancy }
   * @returns {number} shared occupancy
   */
  static getOverlapOccupancy(subset, overlaps) {
    if (subset.length < 2) return 0;
    const sortedSubset = [...subset].sort();
    
    // Find overlap matching exactly this subset
    const match = overlaps.find(o => {
      const sortedO = [...o.zone_ids].sort();
      return (
        sortedO.length === sortedSubset.length &&
        sortedO.every((id, idx) => id === sortedSubset[idx])
      );
    });

    return match ? (match.shared_occupancy || 0) : 0;
  }

  /**
   * Compute true occupancy across given zones using PIE
   * @param {Array} zones - array of zone objects { id, name, current_occupancy, ... }
   * @param {Array} overlaps - array of overlap objects { id, zone_ids, shared_occupancy }
   * @returns {Object} Full breakdown of calculation, terms, raw sum, corrected occupancy, and error
   */
  static calculate(zones, overlaps = []) {
    if (!zones || zones.length === 0) {
      return {
        rawSum: 0,
        correctedOccupancy: 0,
        overcountingError: 0,
        errorPercentage: 0,
        terms: [],
        stepByStepFormula: "Total = 0",
        breakdownHtml: "No active zones selected.",
        setSizes: {},
        intersections: []
      };
    }

    const n = zones.length;
    const zoneMap = new Map(zones.map(z => [z.id, z]));
    const zoneIds = zones.map(z => z.id);

    // 1. Raw sum (Level 1 terms)
    let rawSum = 0;
    const level1Terms = [];
    const setSizes = {};

    zones.forEach(z => {
      const count = Number(z.current_occupancy) || 0;
      rawSum += count;
      setSizes[z.id] = { name: z.name, count };
      level1Terms.push({
        zoneId: z.id,
        name: z.name,
        count
      });
    });

    // 2. Inclusion-Exclusion general expansion
    let correctedOccupancy = 0;
    const termsByLevel = [];
    const intersections = [];
    const formulaParts = [];

    // Level 1: Individual Sets (k=1, positive)
    let level1Total = rawSum;
    correctedOccupancy += level1Total;
    termsByLevel.push({
      k: 1,
      sign: '+',
      title: "Individual Zone Occupancies (Raw Sum)",
      subtotal: level1Total,
      items: level1Terms.map(t => `|${t.name}| = ${t.count}`)
    });
    formulaParts.push(level1Terms.map(t => `|${t.name}|`).join(' + '));

    // Levels 2 to n: Intersections
    let totalSubtracted = 0;
    let totalAddedBack = 0;

    for (let k = 2; k <= n; k++) {
      const combinations = PIECalculator.getCombinations(zoneIds, k);
      const sign = (k % 2 === 1) ? '+' : '−';
      let levelTotal = 0;
      const levelItems = [];

      combinations.forEach(combo => {
        const shared = PIECalculator.getOverlapOccupancy(combo, overlaps);
        if (shared > 0) {
          const names = combo.map(id => zoneMap.get(id)?.name || id);
          const termLabel = `|${names.join(' ∩ ')}|`;
          levelTotal += shared;
          levelItems.push({
            zones: combo,
            names,
            label: termLabel,
            count: shared
          });

          intersections.push({
            k,
            zones: combo,
            names,
            label: termLabel,
            count: shared,
            sign
          });
        }
      });

      if (levelItems.length > 0) {
        if (sign === '−') {
          correctedOccupancy -= levelTotal;
          totalSubtracted += levelTotal;
          formulaParts.push(`− (${levelItems.map(item => item.label).join(' + ')})`);
        } else {
          correctedOccupancy += levelTotal;
          totalAddedBack += levelTotal;
          formulaParts.push(`+ (${levelItems.map(item => item.label).join(' + ')})`);
        }

        termsByLevel.push({
          k,
          sign,
          title: k === 2 
            ? "Pairwise Overlaps (Subtracted to eliminate double-counting)" 
            : `${k}-Way Overlaps (Added back to correct over-subtraction)`,
          subtotal: levelTotal,
          items: levelItems.map(item => `${item.label} = ${item.count}`)
        });
      }
    }

    // Guard against negative occupancy from synthetic bad test inputs
    correctedOccupancy = Math.max(0, correctedOccupancy);

    const overcountingError = rawSum - correctedOccupancy;
    const errorPercentage = rawSum > 0 
      ? Number(((overcountingError / rawSum) * 100).toFixed(1)) 
      : 0;

    const stepByStepFormula = formulaParts.join(' ');

    return {
      rawSum,
      correctedOccupancy,
      overcountingError,
      errorPercentage,
      totalSubtracted,
      totalAddedBack,
      termsByLevel,
      intersections,
      stepByStepFormula,
      setSizes
    };
  }
}

module.exports = PIECalculator;
